### ISSUE TITLE
Build / Use Instructions?

### ISSUE THREAD

**Issue by henrygab** [TricksterGuy/PicrossPuzzleExporter#2]
This looks really interesting!

Can you add to the readme steps to build/use, such as what dependencies there are, what OS are supported, what UI framework is used?

If you define the dependencies in [a way that github understands](https://help.github.com/articles/listing-the-packages-that-a-repository-depends-on), that would be awesome!

**Comment by TricksterGuy** [comment]
Hey thanks for noticing this! This is just a WIP, however I'm just about to make my first release in the coming days.

**Comment by TricksterGuy** [comment]
Also I should note before disappointment sets in...

That this repo only **generates** picross puzzles, it is not a picross game, just an editor for one.

I'm using this program to make a picross game though with all of the variants I came up with in my editor, I'm targeting the Nintendo switch only at first.

**Comment by henrygab** [comment]
Hi @TricksterGuy ,

Yes, I realize it's an editor / creator.  If I read it correctly, it appears to have the following abilities:

1. Import of images
2. Automatic color space reduction from image import
3. Validation of unique solution  (?? does it indicate unsolvable portion ??)
4. Interactive editing (so can minimally modify to make unsolvable into solvable)

If true, that's AWESOME!


Would you be open to releasing under MIT or Apache license?  (If yes, I can offer PRs...)

**Comment by TricksterGuy** [comment]
1. Yes it only imports images, you can't make a puzzle from scratch, or just edit the row/column numbers freely. You have to create an image with the puzzle and then import it.
2. Yes, in all modes other than the "RBY-BW" (I've already renamed them all). the color at that pixel will be used to determine the value at that cell. In RBY-BW I simply convert the image into 8 bit color mode with a predefined palette of the 32 possible colors in that mode.
3. Yes it does validate that there's a unique solution. It was some time ago so I forget the details of how its done.

No it doesn't indicate unsolvability, it is impossible with the way the program is designed to import an unsolvable puzzle since the puzzle is generated from the image. One can after the image is import fix the validation by setting pixels in the editor.

4. Yup interactive editing to fix non-uniqueness errors, the solution hints on the borders will update to reflect changes (though switching modes will erase everything).

Hope this answers all of your questions, and sure! I can reconsider the license.  

ALTHOUGH I'm interested in hearing how you found this little repo and your plans, I'm generally surprised someone else is interested in my made up variations of picross. What PRs did you have in mind as well? The code is pretty unclean and was quickly written in 2014 one weekend where I was bored pretty much so there's some things that could be made more clear/concise.

An additional note that the save format is a [protobuf](https://developers.google.com/protocol-buffers)

Also FYI here's what the software looks like now with my uncommitted changes.
[img](https://photos.app.goo.gl/e3MJQL5fpAJ8hDKm9)

**Comment by henrygab** [comment]
I was looking for a (non-java) nonogram editor / creator.

I searched github and found this one.  The general layout (separation of tasks, logically breaking things into manageable work chunks, etc.) seems grounded.  I admit I don't grok the graphics differencing algorithm (yet), but the project's code appears to have a solid foundation.

I've never used protobuf, so have no opinion on it ... but also don't know its limits / weak points.  Regardless, if needing to add another output format, it seems it would not be too difficult.

PRs would include changing magic constants to be `#define`'d values or `enum` types, aligning protobuf definition of puzzle type (so don't have to +1 / -1), ... nothing major immediately, as I can't even build it yet (which might be me, but I'm not new to coding...).   :smiley:

Speaking of building it ... what libraries is this built on / rely on?  The pic looks promising.  Can this be built to run under Windows, or is it *nix only?

**Comment by TricksterGuy** [comment]
Ahh ok.

Yeah again the code is a mess, just discovered a set of unused files (maybe I had an idea but went a different direction...). It is an embarrassment, but hey weekend project.

I only used protobuf as I didn't want to bother writing a serializer/deserializer which you get for free, plus I use them all the time at work so I already know the API. I'm sure there are converters out there to convert protobuf to say json or xml.

As far as building goes yes it builds on both windows and linux, however my windows build environment is not Visual Studio. I'm using MSYS2/MinGW.  I don't know if this will compile when used in Visual Studio.

So I installed all of the dependencies via MSYS2's package manager.

The dependencies are should be (looking at the CMakeLists.txt)

wxWidgets
protobuf
qrencode

anywho let me clean things up a bit more then you can have at it and also a reminder for me to change the license

**Comment by henrygab** [comment]
Hmmm... If you have the chance, can you help me understand the packages that would need to be installed?  From Ubuntu, I ran the following command, but each search listed many entries:
```bash
for val in wx protobuf qrencode; do apt list *$val*; done
```

As you can see, my best guesses out of the matches doesn't make me hopeful.  😔 

<details><summary>wx</summary><P/>

```
libwxbase3.0-0v5/focal            3.0.4+dfsg-15build1  amd64
libwxbase3.0-dev/focal            3.0.4+dfsg-15build1  amd64
libwxgtk-media3.0-gtk3-0v5/focal  3.0.4+dfsg-15build1  amd64
libwxgtk-media3.0-gtk3-dev/focal  3.0.4+dfsg-15build1  amd64
libwxgtk3.0-gtk3-0v5/focal        3.0.4+dfsg-15build1  amd64
libwxgtk3.0-gtk3-dev/focal        3.0.4+dfsg-15build1  amd64
wx-common/focal                   3.0.4+dfsg-15build1  amd64
wx3.0-headers/focal               3.0.4+dfsg-15build1  all
```
</details><HR/>
<details><summary>protobuf</summary><P/>

```
libprotobuf-c-dev/focal    1.3.3-1           amd64
libprotobuf-dev/focal      3.6.1.3-2ubuntu5  amd64
protobuf-c-compiler/focal  1.3.3-1           amd64
protobuf-compiler/focal    3.6.1.3-2ubuntu5  amd64
```
</details><HR/>
<details><summary>qrencode</summary><P/>

```
libqrencode-dev/focal  4.0.2-2 amd64
libqrencode4/focal     4.0.2-2 amd64
qrencode/focal         4.0.2-2 amd64
```
</details><HR/>

**Comment by TricksterGuy** [comment]
Yeah when it comes to stuff like this always go with the package names ending in -dev

-dev packages include the stuff needed to compile and link applications against said library.
Non dev packages are what you would install if you wanted to run said application.

So

libqrencode-dev libprotobuf-dev libwxgtk3.0-gtk3-dev

not sure about that last one I'd install libwxbase3.0-dev as well if CMake complains.

**Comment by henrygab** [comment]
Generally speaking, install the dependencies, run CMake, manually workaround for protobuf compiler, then run make.

<HR/><details><summary>Ubuntu / Debian Script</summary><P/>

```bash
# Install pre-requisites
for val in cmake libqrencode-dev libprotobuf-dev libwxgtk3.0-gtk3-dev pkg-config libgtk2.0-dev protobuf-compiler; do sudo apt install $val; done

# Configure makefile
cmake .

# Workaround: Regenerate protobuf file
protoc --cpp_out=./src/gen ./gen/Picross.proto

# Build at last!
make clean
make all
```
</details>

<hr/><details><summary>Workaround: Regenerate Protobuf</summary><P/>

Without this step, `make` may fail with the following error:

```bash
[...]/PicrossPuzzleExporter/src/gen/Picross.pb.h:10:10:
   fatal error: google/protobuf/port_def.inc: No such file or directory
   10 | #include <google/protobuf/port_def.inc>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
compilation terminated.
```

This is apparently due to different protobuf compilers generating non-compatible generated source files.  For example, the checked-in files were generated with protobuf compiler version 3.12.2, while Ubuntu may default to version 3.6.1.3, which causes the above error.
</details>

<hr/>

**Comment by TricksterGuy** [comment]
Interesting! So sorry for the troubles man, compiling can be a real pain in the butt sometimes.

What version is libprotobuf-dev at? It could very well be a version mismatch. Installing the package on MSYS2 gave me version 3.12.2

If that's the case you may be able to get away with running 

`protoc --cpp_out=../src/gen Picross.proto` in the gen folder.

I'll make a note to remove Picross.pb.cc/h and have Cmake generate that file instead that way should compile regardless of what version libprotobuf-dev is at.

**Comment by henrygab** [comment]
Seems to be an older version **3.6.1.3**:

`libprotobuf-dev/focal,now 3.6.1.3-2ubuntu5 amd64 [installed]`

With the above workaround, I can confirm `make` successfully build under Ubuntu (WSL).  See above post, which I updated with bash script.  Obviously, it doesn't actually **_run_** under WSL ([**_yet_**](https://www.theverge.com/2020/5/19/21263377/microsoft-windows-10-linux-gui-apps-gpu-acceleration-wsl-features)), but hopefully that part will just happen automagically later.

Again, thank you!  Being able to build will at least allow me to sanity-check that any changes I later propose compile. 👍   I'll await your notification that you're done with whatever it is you wanted to cleanup and updated the license, before I make a PR.

Now to see if I can get this to compile / run for Windows, without MSYS2...

**Comment by TricksterGuy** [comment]
Thanks for that and the commit comments. I really just want to go through and rewrite the whole dang program at this point, there's a lot of stuff I've learned about C++ and CMake since 2014 that can be applied here.

I'm playtesting the various modes I have came up with now. My conclusion is that some changes are required to make them work better, there's a lot of trial and error in solving it.

Hadn't really shared *that* openly yet but here's an 8x8 colors of light bpc=2 puzzle if interested.

https://docs.google.com/spreadsheets/d/1li2tdPC1fg3aXiwtLkSB4SoAAwvxvdbrGMtnLHHocdA/edit?usp=sharing

**Comment by henrygab** [comment]
Hmm.... I was able to solve the picros / nonogram portion without trouble.

That allowed me to mark each cell as either "at least one", or "must be zero".

I was able to mark each cell in the row totaling 23 as "at least two", knowing that all values must be 3, except for a single value of 2.

Beyond this, I reached a pretty steep difficulty curve.  After about ten minutes, I could not even be sure that local progress can be made, without guessing.

After further review, I realize the current code does not validate the puzzle.  (guarantee a **unique** solution).  Generally, having a unique solution is considered a requirement by puzzle lovers.


I love the possibility of the multi-layer picross, which are solved independently, but merge (one layer maps to one RGB value, result is sum of all layers' R, G, and B values) for the final image.  The beauty of this is many:

1. Puzzle solvers don't need to learn new solve methods
2. Puzzle solvers can solve one layer, without it revealing the final image
3. [Many libraries exist](https://webpbn.com/survey/#simpson) for ensuring the solution of each layer is unique (e.g., [pbnsolve](http://code.google.com/p/pbnsolve/source/browse/) is Apache 2.0, [gecode](https://github.com/Gecode/gecode) is MIT)

In short, you get to focus on UI/interactivity, while the validation problem has already been solved, and could then bring this new variant of nonogram out more quickly.  By quickly validating the solution is unique, it can avoid user frustration.

pbnsolve also has a number of very useful features for determining how difficult it would be for a human to solve the puzzle.


I think I will defer further work in this space for now.  Thanks for the great interactions!

**Comment by TricksterGuy** [comment]
Yup thanks for the feedback!

As I probably mentioned several times (feel like I'm a broken record). I really hadn't taken this project seriously. I use such things to implement an interesting algorithm on my own probably because work that week wasn't challenging enough for me. Nice to haves would be your standard rigor behind a codebase, unit tests and all I just used my time to invest in getting something working for my purposes.

I will take those suggestions to heart, I don't even remember what my algorithm for uniqueness doesn't exactly, but integration with a library that does it better seems desirable.

New version of that sheet is here if you are still interested:  I've changed the format https://docs.google.com/spreadsheets/d/11vCGLi0XlpDunlhda6xYhH6PJgNCSxbhD6QUzX022DA/edit?usp=sharing
